File: 2020/day25/day25.kt
fun main() {
    val M = 20201227
    fun encode(s: Int) = generateSequence(1L) { it * s % M }
    fun transform(s: Int, n: Int) = encode(s).take(n + 1).last()
    fun reverse(x: Int) = encode(7).withIndex()
        .takeWhile { (_, v) -> v.toInt() != x }.last().index + 1

    val (A, B) = Pair(10705932, 12301431)
    println(transform(B, reverse(A)))
    println(transform(A, reverse(B)))
}

File: 2020/day25/day25.py
N, M = 7, 20201227

def transform(subj, loop):
    n = 1
    for _ in range(loop):
        n = n * subj % M
    return n

def reverse(res):
    n, s = 1, 0
    while n != res:
        n, s = n * N % M, s + 1
    return s

A, B = 8987316, 14681524
print(transform(B, reverse(A)))
print(transform(A, reverse(B)))
